### PR TITLE
Blog app - integration specs for views and fix n+1 problems

### PR DESCRIPTION
**Pull request by DianaBeki**
### [Project requirements](https://github.com/microverseinc/curriculum-rails/blob/main/blog-app/projects/intergration_spec_and_n_1_project.md#project-requirements)
[n+1](https://github.com/microverseinc/curriculum-rails/blob/main/blog-app/projects/intergration_spec_and_n_1_project.md#n1)
- Here are  the screenshots of my app's logs before and after the fix
after
after1
before
before1
userpost
includes

This is the corresponding wireframe view: blog all user posts
[Integration specs](https://github.com/microverseinc/curriculum-rails/blob/main/blog-app/projects/intergration_spec_and_n_1_project.md#integration-specs)

- User index page:

    - I can see the username of all other users.
    - I can see the profile picture for each user.
   - I can see the number of posts each user has written.
   - When I click on a user, I am redirected to that user's show page.
   - user show page:

- I can see the user's profile picture.
- I can see the user's username.
- I can see the number of posts the user has written.
- I can see the user's bio.
- I can see the user's first 3 posts.
- I can see a button that lets me view all of a user's posts.
- When I click a user's post, it redirects me to that post's show page.
- When I click to see all posts, it redirects me to the user's post's index page.
- User post index page:

- I can see the user's profile picture.
- I can see the user's username.
- I can see the number of posts the user has written.
- I can see a post's title.
- I can see some of the post's body.
- I can see the first comments on a post.
- I can see how many comments a post has.
- I can see how many likes a post has.
- I can see a section for pagination if there are more posts than fit on the view.
- When I click on a post, it redirects me to that post's show page.
- Post show page:
- 
- I can see the post's title.
- I can see who wrote the post.
- I can see how many comments it has.
- I can see how many likes it has.
- I can see the post body.
- I can see the username of each commentor.
- I can see the comment each commentor left.
- Be sure to include integration specs for any other features you may have implemented!

### CODE REVIEW

File: spec/features/user_show_spec.rb [hunk 1]
```diff
@@ -0,0 +1,63 @@
...
+                    likes_counter: 0),
+        Post.create(author: @diana, title: 'User Post 2', text: 'This is User Post 2', comments_counter: 0,
+                    likes_counter: 0),
+        Post.create(author: @diana, title: 'User Post 3', text: 'This is User Post 3', comments_counter: 0,
+                    likes_counter: 0)
+      ]
+      visit user_path(@diana)
+    end
+    it "can see the user's profile picture" do
+      #  expect(page).to have_css("img[src*='#{@diana.photo}']")
```
Comment (t-yanick) [inline spec/features/user_show_spec.rb]: - [ ] Great job working on this milestone. In order to keep out codebase clean and professional, we should always delete all commented lines of code.

File: spec/features/user_show_spec.rb [hunk 2]
```diff
@@ -0,0 +1,63 @@
...
+    end
+
+    it "can see the user's first 3 posts" do
+      @diana.posts.limit(3).each do |post|
+        expect(page).to have_content(post.text)
+      end
+    end
+
+    it "can see a button that lets me view all of a user's posts" do
+      # expect(page).to have_link('See All Posts', href: user_posts_path(@diana), class: 'btn center-btn')
```
Comment (t-yanick) [inline spec/features/user_show_spec.rb]: - [ ] Great job working on this milestone. In order to keep out codebase clean and professional, we should always delete all commented lines of code.

File: spec/features/post_index_spec.rb [hunk 3]
```diff
@@ -0,0 +1,82 @@
...
+    end
+
+    it 'can see the number of posts the user has written.' do
+      expect(page).to have_content(@user.posts_counter.to_s)
+    end
+
+    it 'can see posts title' do
+      # @posts.each do |post|
+      #   expect(page).to have_content(post.title.to_s)
+      # end
```
Comment (t-yanick) [inline spec/features/post_index_spec.rb]: - [ ] Great job working on this milestone. In order to keep out codebase clean and professional, we should always delete all commented lines of code.

File: spec/features/post_index_spec.rb [hunk 4]
```diff
@@ -0,0 +1,82 @@
...
+      expect(page).to have_content 'User Post 1'
+    end
+
+    it 'can see some of the body of the post' do
+      # @posts.each do |post|
+      #   expect(page).to have_content(post.text.to_s)
+      # end
```
Comment (t-yanick) [inline spec/features/post_index_spec.rb]: - [ ] Great job working on this milestone. In order to keep out codebase clean and professional, we should always delete all commented lines of code.

File: spec/features/post_index_spec.rb [hunk 5]
```diff
@@ -0,0 +1,82 @@
...
+      end
+    end
+
+    it 'can see how many likes a post has' do
+      @posts.each do |post|
+        expect(page).to have_content(post.likes_counter.to_s)
+      end
+    end
+  end
+
```
Comment (t-yanick) [inline spec/features/post_index_spec.rb:72]: - [ ] Great job adding most of the test cases. You are almost there. At the moment, you are missing the test that asserts you can see a section for pagination if there are more posts than fit on the view. Kindly add this test case.

File: spec/features/post_show_spec.rb [hunk 6]
```diff
@@ -0,0 +1,52 @@
...
+    @posts = [
+      @post1 = Post.create(author: @user, title: 'User Post 1', text: 'This is User Post 1', likes_counter: 0,
+                           comments_counter: 0),
+      @post2 = Post.create(author: @user, title: 'User Post 2', text: 'This is User Post 2', likes_counter: 0,
+                           comments_counter: 0),
+      @post3 = Post.create(author: @user, title: 'User Post 3', text: 'This is User Post 3', likes_counter: 0,
+                           comments_counter: 0)
+    ]
+
+    # visit user_posts_url(user_id: @user.id)
```
Comment (t-yanick) [inline spec/features/post_show_spec.rb]: - [ ] Great job working on this milestone. In order to keep out codebase clean and professional, we should always delete all commented lines of code.

**Review by t-yanick — CHANGES_REQUESTED** [review]
# STATUS: CHANGES REQUIRED

Hi @DianaBeki , 

### Good job so far!
### To Highlight! :clap:  :green_circle: 

:heavy_check_mark:  Collaborative work 👍 Commits from both team members
:heavy_check_mark: Screenshots of solution to N + 1 problem added to PR body
:heavy_check_mark: Effective use of testing gem
:heavy_check_mark: Good commit messages
:heavy_check_mark: Use of correct gitflow

There are some issues that you still need to work on to go to the next project but you are almost there!

## Required Changes ♻️
_Check the comments under the review._

## Optional suggestions
_Every comment with the [OPTIONAL] prefix is not crucial enough to stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better._

You can also consider:
- N/A

Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question so I can receive the notification.

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561) you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

**Review by Gambit142 — APPROVED** [review]
Hi @DianaBeki @yin-ka @rubydevi 👋,

## Status: Approved 🟢   

Your project is complete!  There is nothing else to say other than... it's time to merge it :shipit: 
Congratulations! 🎉

## Highlights
- Setup linters properly ✔️ 
- Used `Rspec` as testing library ✔️ 
- Installed `capybara` gem ✔️
- Wrote feature tests for Users and Posts pages ✔️  
- Solved **_N+1 query_** problem ✔️ 

Cheers, and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.

#### Please tag me in your question so I can receive the notification. You can tag me by typing @Gambit142

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561)  you have a limited number of reviews per project (check the exact number in your Dashboard). If you think the code review was unfair, request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

**Pull request merged**